### PR TITLE
Export metadata

### PR DESCRIPTION
**Pull request by benthorner**
https://trello.com/c/7k1kBrqc/170-add-timestamp-as-metadata-to-the-contentjsongz-and-taxonsjsongz-files

### CODE REVIEW

File: python/data_extraction/export_data.py [hunk 1]
```diff
@@ -145,3 +87,68 @@ def __taxonomy():
...
+    content_links_set = set(content_links_list)
+    duplicate_links = len(content_links_list) - len(content_links_set)
+
+    if duplicate_links > 0:
+        print("{} duplicate links from Rummager".format(duplicate_links))
+
+    pool = Pool(10)
+    return pool.imap(get_content, content_links_set), len(content_links_set)
+
+
```
Comment (cbaines) [inline python/data_extraction/export_data.py]: It's difficult to see the changes here, as all the code has moved. Would it be possible to do this in a separate commit?

File: python/data_extraction/export_data.py [hunk 2]
```diff
@@ -1,12 +1,16 @@
-import data
 from data_extraction import content_export
 from data_extraction import taxonomy_query
 from lib import plek
 from lib.helpers import dig
+from multiprocessing import Pool
```
Comment (cbaines) [inline python/data_extraction/export_data.py]: This doesn't seem to be used in this commit. If it is used later, maybe move it there?

**Comment by koetsier** [comment]
Maybe add a description/link to trello

File: python/test/data_extraction/test_export_data.py [hunk 3]
```diff
@@ -65,4 +66,16 @@ def test_export_untagged_content(self):
...
+    @freeze_time("2018-03-07")
+    def test_export_metadata(self):
+        output = MockIO()
+        input_string = json.dumps([content_with_taxons, content_without_taxons])
+
+        with unittest.mock.patch('gzip.open',
+                                 side_effect=self.return_input_or_output(io.StringIO(input_string), output)):
+            export_data.export_untagged_content()
+            expected = {"date": "2018-03-07 00:00:00", 
+                        "code": Repo(search_parent_directories=True).head.commit.hexsha}
```
Comment (koetsier) [inline python/test/data_extraction/test_export_data.py]: Can Repo be mocked?

Comment (benthorner) [reply python/test/data_extraction/test_export_data.py]: I looked and couldn't see a way - more symptomatic of the script having internal dependencies, which could be passed-in. Think that's going a bit far, though.

File: python/data_extraction/export_data.py (hunk 2 above)
Comment (benthorner) [reply python/data_extraction/export_data.py]: The diff makes it hard to see that I reordered / separated from ... import vs. import.

File: python/data_extraction/export_data.py [hunk 4]
```diff
@@ -105,7 +109,9 @@ def __bool__(self):
         def __iter__(self):
             return self.iterator
 
-    obj = { "items": StreamContent(iterator) }
+    obj = {"_meta": {"date": str(datetime.datetime.today()),
+                     "code": Repo(search_parent_directories=True).head.commit.hexsha},
+           "items": StreamContent(iterator)}
```
Comment (cbaines) [inline python/data_extraction/export_data.py]: My preference would be to indent this like:
```
    obj = {
        "_meta": {
            "date": str(datetime.datetime.today()),
            "code": Repo(search_parent_directories=True).head.commit.hexsha
        },
        "items": StreamContent(iterator)
}
```

Mostly because this is how I'm used to doing it, but also because the lines will be shorter.

File: python/requirements.txt [hunk 5]
```diff
@@ -18,3 +18,4 @@ statsd==3.2.2
 progressbar2==3.35.2
 ijson==2.3
 gitpython==2.1.8
+freezegun==0.3.10
```
Comment (cbaines) [inline python/requirements.txt]: I think freezegun is being used in the previous commit, and it would be good to add the requirement there also.

File: python/test/data_extraction/test_export_data.py [hunk 6]
```diff
@@ -50,14 +53,16 @@ def test_export_filtered_content(self):
...
         output = MockIO()
         input_string = json.dumps([content_with_taxons, content_without_taxons])
 
         with unittest.mock.patch('gzip.open',
                                  side_effect=self.return_input_or_output(io.StringIO(input_string), output)):
-            export_data.export_filtered_content()
-            self.assertListEqual([content_without_taxons], json.loads(output.buffer)["items"])
+            export_data.export_untagged_content()
+            expected = [{"title":"title1"}]
+            self.assertListEqual(expected, json.loads(output.buffer)["items"])
```
Comment (cbaines) [inline python/test/data_extraction/test_export_data.py]: There is a missing trailing newline here.

**Comment by cbaines** [comment]
This looks really good, let's just tidy up the commits and trailing newlines before merging.

**Comment by koetsier** [comment]
Can you write a description in the pull request and a link to the trello ticket?

File: python/test/data_extraction/test_export_data.py (hunk 3 above)
Comment (koetsier) [reply python/test/data_extraction/test_export_data.py]: Could 'patch' Repo - something similar is happening in this file when patching gzip.open

**Comment by benthorner** [comment]
https://trello.com/c/7k1kBrqc/170-add-timestamp-as-metadata-to-the-contentjsongz-and-taxonsjsongz-files

File: python/test/data_extraction/test_export_data.py (hunk 3 above)
Comment (benthorner) [reply python/test/data_extraction/test_export_data.py]: I think that only applies to a single method invocation. The Repo bit is actually a constructor - I don't think you can override those.

Comment (koetsier) [reply python/test/data_extraction/test_export_data.py]: You could use:
```
    @patch('data_extraction.export_data.Repo')
    def test_export_metadata(self, MockedRepo):
        MockedRepo.return_value = Mock(commit=Mock(hexsha='myhexsha'))
```

Comment (benthorner) [reply python/test/data_extraction/test_export_data.py]: That works - thanks Jos!

File: python/data_extraction/export_data.py [hunk 7]
```diff
@@ -145,3 +91,68 @@ def __taxonomy():
...
+    )
+
+    content_links_set = set(content_links_list)
+    duplicate_links = len(content_links_list) - len(content_links_set)
+
+    if duplicate_links > 0:
+        print("{} duplicate links from Rummager".format(duplicate_links))
+
+    pool = Pool(10)
+    return pool.imap(get_content, content_links_set), len(content_links_set)
```
Comment (cbaines) [inline python/data_extraction/export_data.py]: There is a missing trailing newline here.

File: .gitignore [hunk 8]
```diff
@@ -8,6 +8,7 @@ checkpoints
 data
 .DS_Store
 __pycache__
-*/.cache
 client_secret.json
 python/notebooks/ML_notebooks/Graph/
+**/.cache
+*.log
```
Comment (cbaines) [inline .gitignore]: There is a missing trailing newline here.

File: python/data_extraction/export_data.py (hunk 4 above)
Comment (benthorner) [reply python/data_extraction/export_data.py]: I'm happy to do either way. Should we add a linter to help with this sort of thing?

Comment (cbaines) [reply python/data_extraction/export_data.py]: Yes, adding some kind of linting would be good (but probably in a new Pull Request).
